src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,532 @@
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_tcp_bind(tcp_socket->m_uv_tcp, (const struct sockaddr *)&addr_ptr, 0);
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        std::string err = std::string("failed to bind tcp_socket: ") + uv_strerror(result);
+        return io_result_mk_error(err.c_str());
+    }
+
+    return lean_io_result_mk_ok(lean_box(0));
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_listen(b_obj_arg socket, int32_t backlog) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    event_loop_lock(&global_ev);
+
+    // The event loop owns the socket.
+    lean_inc(socket);
+
+    int result = uv_listen((uv_stream_t*)tcp_socket->m_uv_tcp, backlog, [](uv_stream_t* stream, int status) {
+        lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket((lean_object*)stream->data);
+
+        if (tcp_socket->m_promise_accept == NULL) {
+            return;
+        }
+
+        lean_object * promise = tcp_socket->m_promise_accept;
+
+        if (status < 0) {
+            resolve_promise_with_status(promise, status);
+            return;
+        }
+
+        lean_object * client = unpack_io(lean_uv_tcp_new());
+        lean_uv_tcp_socket_object * client_socket = lean_to_uv_tcp_socket(client);
+
+        int result = uv_accept((uv_stream_t*)tcp_socket->m_uv_tcp, (uv_stream_t*)client_socket->m_uv_tcp);
+
+        if (result < 0) {
+            lean_dec(client);
+            resolve_promise_with_status(promise, result);
+            return;
+        }
+
+        tcp_socket->m_promise_accept = NULL;
+
+        resolve_promise(promise, mk_ok_except(client));
+    });
+
+    if (result < 0) {
+        std::string err = std::string("failed to listen tcp_socket: ") + uv_strerror(result);
+        return io_result_mk_error(err.c_str());
+    }
+
+    event_loop_unlock(&global_ev);
+
+    return lean_io_result_mk_ok(lean_box(0));
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_accept(b_obj_arg socket) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    if (tcp_socket->m_promise_accept != NULL) {
+        std::string err = std::string("failed to accept tcp_socket: parallel accept is not allowed!");
+        return io_result_mk_error(err.c_str());
+    }
+
+    lean_object * promise = create_promise();
+    tcp_socket->m_promise_accept = promise;
+
+    lean_object * client = unpack_io(lean_uv_tcp_new());
+
+    lean_uv_tcp_socket_object * client_socket = lean_to_uv_tcp_socket(client);
+
+    int result = uv_accept((uv_stream_t*)tcp_socket->m_uv_tcp, (uv_stream_t*)client_socket->m_uv_tcp);
+
+    if (result < 0 && result != UV_EAGAIN) {
+        lean_dec(client);
+        resolve_promise_with_status(promise, result);
+    } else if (result >= 0) {
+        resolve_promise(promise, mk_ok_except(client));
+        tcp_socket->m_promise_accept = NULL;
+    } else {
+        lean_dec(client);
```

Comment:
As a little optimization note: While it is correct that we can `lean_dec` the client here we could *theoretically* also cache the client in TCP server struct to avoid allocating it twice though that is something we can safe up for later.

Reply:
Implemented!